File: packages/procm-sdk/src/hook.ts
import callsites from "callsites";
import type { ProcmClient } from "./client.js";
import type { JsonValue } from "./protocol.js";
import { saveTrace } from "./trace.js";

export interface TraceFrame {
  index: number;
  functionName: string;
  file: string;
  line: number | null;
  column: number | null;
  async: boolean;
}

export interface FunctionTrace {
  kind: "function";
  traceId: string;
  name: string;
  startedAt: number;
  durationMs: number;
  status: "returned" | "resolved" | "threw" | "rejected" | "skipped";
  callChain: TraceFrame[];
  args?: JsonValue;
  result?: JsonValue;
  error?: { name: string; message: string; stack?: string };
}

export interface CreateHookOptions {
  client?: ProcmClient;
  name?: string;
  captureArgs?: boolean;
  captureResult?: boolean;
  ttlSeconds?: number;
  filterFrame?: (frame: TraceFrame) => boolean;
  onTraceCreated?: (traceId: string) => void;
  onStored?: (traceId: string) => void;
  onStoreError?: (error: Error, traceId: string) => void;
}

export interface BeforeHookContext {
  traceId: string;
  args: unknown[];
  callChain: TraceFrame[];
  setArgs(args: unknown[]): void;
  skip(result?: unknown): void;
}

export interface AfterHookContext {
  traceId: string;
  args: unknown[];
  result: unknown;
  error?: unknown;
  callChain: TraceFrame[];
  setResult(result: unknown): void;
}

export type BeforeHookHandler = (context: BeforeHookContext) => void;
export type AfterHookHandler = (context: AfterHookContext) => void;

export type HookedFunction<TFunction extends (...args: any[]) => any> = TFunction & {
  before(handler: BeforeHookHandler): HookedFunction<TFunction>;
  after(handler: AfterHookHandler): HookedFunction<TFunction>;
  readonly original: TFunction;
};

export interface PropertyHookOptions extends CreateHookOptions {
  captureGet?: boolean;
  captureSet?: boolean;
}

function randomId(): string {
  return globalThis.crypto?.randomUUID?.() ?? `${Date.now().toString(36)}-${Math.random().toString(36).slice(2)}`;
}

function captureCallChain(filterFrame?: (frame: TraceFrame) => boolean): TraceFrame[] {
  const frames = callsites()
    .map((site, index): TraceFrame => ({
      index,
      functionName: site.getFunctionName() ?? site.getMethodName() ?? "<anonymous>",
      file: site.getFileName() ?? "<unknown>",
      line: site.getLineNumber(),
      column: site.getColumnNumber(),
      async: site.isAsync(),
    }))
    .filter((frame) => !frame.file.endsWith("/hook.js") && !frame.file.endsWith("/hook.ts"))
    .filter((frame) => filterFrame?.(frame) ?? true)
    .slice(0, 100)
    .map((frame, index) => ({ ...frame, index }));
  return frames;
}

function jsonValueOrPlaceholder(value: unknown): JsonValue {
  try {
    const text = JSON.stringify(value);
    if (text === undefined) return { unavailable: "not JSON serializable" };
    return JSON.parse(text) as JsonValue;
  } catch {
    return { unavailable: "not JSON serializable" };
  }
}

function errorDetails(error: unknown): FunctionTrace["error"] {
  if (error instanceof Error) return { name: error.name, message: error.message, stack: error.stack };
  return { name: "Error", message: String(error) };
}

function assertSynchronous(result: unknown, phase: "before" | "after"): void {
  if (result !== null && (typeof result === "object" || typeof result === "function") &&
      typeof (result as { then?: unknown }).then === "function") {
    throw new Error(`${phase} hook handlers must be synchronous`);
  }
}

function store(options: CreateHookOptions, trace: FunctionTrace): void {
  if (!options.client) return;
  void saveTrace(options.client, trace as unknown as JsonValue, { id: trace.traceId, ttlSeconds: options.ttlSeconds })
    .then((id) => options.onStored?.(id))
    .catch((error: unknown) => options.onStoreError?.(error instanceof Error ? error : new Error(String(error)), trace.traceId));
}

export function createHook<TFunction extends (...args: any[]) => any>(
  fn: TFunction,
  options: CreateHookOptions = {},
): HookedFunction<TFunction> {
  if (typeof fn !== "function") throw new TypeError("createHook requires a function");
  const beforeHandlers: BeforeHookHandler[] = [];
  const afterHandlers: AfterHookHandler[] = [];

  const hooked = function (this: unknown, ...initialArgs: unknown[]) {
    const traceId = randomId();
    const startedAt = Date.now();
    const started = performance.now();
    const callChain = captureCallChain(options.filterFrame);
    let args = initialArgs;
    let skipped = false;
    let currentResult: unknown;
    options.onTraceCreated?.(traceId);

    for (const handler of beforeHandlers) {
      const returned = handler({
        traceId,
        args,
        callChain,
        setArgs(next) { args = next; },
        skip(result) { skipped = true; currentResult = result; },
      });
      assertSynchronous(returned, "before");
    }

    const complete = (
      status: FunctionTrace["status"],
      result: unknown,
      error?: unknown,
    ): unknown => {
      currentResult = result;
      for (const handler of afterHandlers) {
        const returned = handler({
          traceId,
          args,
          result: currentResult,
          error,
          callChain,
          setResult(next) { currentResult = next; },
        });
        assertSynchronous(returned, "after");
      }
      const trace: FunctionTrace = {
        kind: "function",
        traceId,
        name: options.name ?? fn.name ?? "<anonymous>",
        startedAt,
        durationMs: performance.now() - started,
        status,
        callChain,
      };
      if (options.captureArgs) trace.args = jsonValueOrPlaceholder(args);
      if (options.captureResult && error === undefined) trace.result = jsonValueOrPlaceholder(currentResult);
      if (error !== undefined) trace.error = errorDetails(error);
      store(options, trace);
      return currentResult;
    };

    if (skipped) return complete("skipped", currentResult);

    let result: ReturnType<TFunction>;
    try {
      result = fn.apply(this, args as Parameters<TFunction>);
    } catch (error) {
      complete("threw", undefined, error);
      throw error;
    }
    if (result !== null && (typeof result === "object" || typeof result === "function") &&
        typeof (result as { then?: unknown }).then === "function") {
      return Promise.resolve(result).then(
        (value) => complete("resolved", value),
        (error) => {
          complete("rejected", undefined, error);
          throw error;
        },
      );
    }
    return complete("returned", result);
  } as TFunction;

  Object.defineProperties(hooked, {
    original: { value: fn, enumerable: true },
    before: { value: (handler: BeforeHookHandler) => { beforeHandlers.push(handler); return hooked; } },
    after: { value: (handler: AfterHookHandler) => { afterHandlers.push(handler); return hooked; } },
  });
  return hooked as HookedFunction<TFunction>;
}

export function hookProperty<T extends object, K extends keyof T>(
  target: T,
  key: K,
  options: PropertyHookOptions = {},
): () => void {
  const descriptor = Object.getOwnPropertyDescriptor(target, key);
  if (!descriptor) throw new Error(`property ${String(key)} must be an own property`);
  if (!descriptor.configurable) throw new Error(`property ${String(key)} must be configurable`);

  let restored = false;
  let currentValue = descriptor.value;
  const originalGet = descriptor.get;
  const originalSet = descriptor.set;
  const getValue = function (this: T): unknown {
    return originalGet ? originalGet.call(this) : currentValue;
  };
  const setValue = function (this: T, value: unknown): void {
    if (originalSet) originalSet.call(this, value);
    else if ("writable" in descriptor && descriptor.writable) currentValue = value;
  };
  const hookedGet = options.captureGet === false
    ? getValue
    : createHook(getValue, { ...options, name: options.name ?? `${String(key)}:get`, captureResult: options.captureResult ?? true });
  const hookedSet = options.captureSet === false
    ? setValue
    : createHook(setValue, { ...options, name: options.name ?? `${String(key)}:set`, captureArgs: options.captureArgs ?? true });

  Object.defineProperty(target, key, {
    configurable: true,
    enumerable: descriptor.enumerable,
    get: hookedGet,
    set: originalSet || ("writable" in descriptor && descriptor.writable) ? hookedSet : undefined,
  });

  return () => {
    if (restored) return;
    restored = true;
    const restoredDescriptor = "value" in descriptor ? { ...descriptor, value: currentValue } : descriptor;
    Object.defineProperty(target, key, restoredDescriptor);
  };
}

File: packages/procm-sdk-dart/lib/src/hook.dart
import 'dart:convert';

import 'package:stack_trace/stack_trace.dart';

import 'client.dart';
import 'protocol.dart';
import 'trace.dart';
import 'utils.dart';

class TraceFrame {
  const TraceFrame({
    required this.index,
    required this.functionName,
    required this.file,
    this.line,
    this.column,
    required this.isAsync,
  });

  factory TraceFrame.fromJson(Map<String, Object?> json) => TraceFrame(
        index: (json['index'] as num?)?.toInt() ?? 0,
        functionName: json['functionName'] as String? ?? '<anonymous>',
        file: json['file'] as String? ?? '<unknown>',
        line: (json['line'] as num?)?.toInt(),
        column: (json['column'] as num?)?.toInt(),
        isAsync: json['async'] as bool? ?? false,
      );

  final int index;
  final String functionName;
  final String file;
  final int? line;
  final int? column;

  /// Dart stack traces cannot reliably flag async frames; always false here
  /// (the wire field keeps compatibility with the TS SDK's format).
  final bool isAsync;

  Map<String, Object?> toJson() => {
        'index': index,
        'functionName': functionName,
        'file': file,
        'line': line,
        'column': column,
        'async': isAsync,
      };
}

class FunctionTrace {
  const FunctionTrace({
    required this.kind,
    required this.traceId,
    required this.name,
    required this.startedAt,
    required this.durationMs,
    required this.status,
    required this.callChain,
    this.args,
    this.result,
    this.error,
  });

  factory FunctionTrace.fromJson(Map<String, Object?> json) => FunctionTrace(
        kind: json['kind'] as String? ?? 'function',
        traceId: json['traceId'] as String,
        name: json['name'] as String? ?? '<anonymous>',
        startedAt: (json['startedAt'] as num?)?.toInt() ?? 0,
        durationMs: (json['durationMs'] as num?)?.toDouble() ?? 0,
        status: json['status'] as String? ?? 'returned',
        callChain: (json['callChain'] as List? ?? const [])
            .whereType<Map>()
            .map((frame) => TraceFrame.fromJson(frame.cast<String, Object?>()))
            .toList(),
        args: json['args'],
        result: json['result'],
        error: json['error'] is Map
            ? Map<String, Object?>.from(json['error'] as Map)
            : null,
      );

  final String kind;
  final String traceId;
  final String name;
  final int startedAt;
  final double durationMs;
  final String status; // returned | resolved | threw | rejected | skipped
  final List<TraceFrame> callChain;
  final JsonValue args;
  final JsonValue result;
  final Map<String, Object?>? error;

  Map<String, Object?> toJson() => {
        'kind': kind,
        'traceId': traceId,
        'name': name,
        'startedAt': startedAt,
        'durationMs': durationMs,
        'status': status,
        'callChain': callChain.map((frame) => frame.toJson()).toList(),
        if (args != null) 'args': args,
        if (result != null) 'result': result,
        if (error != null) 'error': error,
      };
}

class CreateHookOptions {
  const CreateHookOptions({
    this.client,
    this.name,
    this.captureArgs = false,
    this.captureResult = false,
    this.ttlSeconds,
    this.filterFrame,
    this.onTraceCreated,
    this.onStored,
    this.onStoreError,
  });

  final ProcmClient? client;
  final String? name;
  final bool captureArgs;
  final bool captureResult;
  final int? ttlSeconds;
  final bool Function(TraceFrame frame)? filterFrame;
  final void Function(String traceId)? onTraceCreated;
  final void Function(String traceId)? onStored;
  final void Function(Object error, String traceId)? onStoreError;
}

class BeforeHookContext {
  BeforeHookContext({
    required this.traceId,
    required this.args,
    required this.callChain,
  });

  final String traceId;
  List<Object?> args;
  final List<TraceFrame> callChain;
  bool skipped = false;
  Object? skipResult;

  void setArgs(List<Object?> next) => args = next;

  void skip([Object? result]) {
    skipped = true;
    skipResult = result;
  }
}

class AfterHookContext {
  AfterHookContext({
    required this.traceId,
    required this.args,
    required this.result,
    this.error,
    required this.callChain,
  });

  final String traceId;
  final List<Object?> args;
  Object? result;
  final Object? error;
  final List<TraceFrame> callChain;

  void setResult(Object? next) => result = next;
}

typedef BeforeHookHandler = void Function(BeforeHookContext context);
typedef AfterHookHandler = void Function(AfterHookContext context);

/// Wraps [fn] with a trace pipeline: call-chain capture, before/after
/// handlers, optional args/result capture and background trace storage.
///
/// Unlike the TS SDK (which returns the hooked function itself), Dart cannot
/// transparently forward arbitrary signatures, so call [apply] (or
/// [applyAs]) and use [asFunction] where a `Function` value is required.
class HookedFunction {
  HookedFunction._(this._fn, this._options);

  final Function _fn;
  final CreateHookOptions _options;
  final List<BeforeHookHandler> _beforeHandlers = [];
  final List<AfterHookHandler> _afterHandlers = [];

  /// The unwrapped original function.
  Function get original => _fn;

  HookedFunction before(BeforeHookHandler handler) {
    _beforeHandlers.add(handler);
    return this;
  }

  HookedFunction after(AfterHookHandler handler) {
    _afterHandlers.add(handler);
    return this;
  }

  /// Typed convenience over [apply].
  R applyAs<R>(List<Object?> positionalArgs) => apply(positionalArgs) as R;

  /// Runs the wrapped function through the hook pipeline.
  Object? apply(List<Object?> positionalArgs) {
    final traceId = randomId();
    final startedAt = DateTime.now().millisecondsSinceEpoch;
    final stopwatch = Stopwatch()..start();
    final callChain = captureCallChain(filterFrame: _options.filterFrame);
    var args = positionalArgs;
    final before = BeforeHookContext(traceId: traceId, args: args, callChain: callChain);
    _options.onTraceCreated?.call(traceId);
    for (final handler in List.of(_beforeHandlers)) {
      handler(before);
    }
    args = before.args;

    Object? complete(String status, Object? result, [Object? error]) {
      var currentResult = result;
      final after = AfterHookContext(
        traceId: traceId,
        args: args,
        result: currentResult,
        error: error,
        callChain: callChain,
      );
      for (final handler in List.of(_afterHandlers)) {
        handler(after);
      }
      currentResult = after.result;
      final trace = FunctionTrace(
        kind: 'function',
        traceId: traceId,
        name: _options.name ?? _functionName(_fn),
        startedAt: startedAt,
        durationMs: stopwatch.elapsedMicroseconds / 1000,
        status: status,
        callChain: callChain,
        args: _options.captureArgs ? _jsonValueOrPlaceholder(args) : null,
        result: _options.captureResult && error == null ? _jsonValueOrPlaceholder(currentResult) : null,
        error: error != null ? _errorDetails(error) : null,
      );
      _store(_options, trace);
      return currentResult;
    }

    if (before.skipped) return complete('skipped', before.skipResult);

    Object? result;
    try {
      result = Function.apply(_fn, args);
    } catch (error) {
      complete('threw', null, error);
      rethrow;
    }
    if (result is Future) {
      return result.then(
        (value) => complete('resolved', value),
        onError: (Object error) {
          complete('rejected', null, error);
          throw error;
        },
      );
    }
    return complete('returned', result);
  }

  /// Returns the hooked pipeline as a plain `Function(List<Object?>)`.
  /// Cast it back to the original signature at the call site:
  /// `hooked.asFunction() as int Function(int)`.
  Function asFunction() => apply;
}

String _functionName(Function fn) {
  final text = fn.toString();
  // e.g. "int Function(int)" (tear-off) or "Closure: () => ...".
  return text.startsWith('Closure: ') ? text.substring(9) : text;
}

List<TraceFrame> captureCallChain({bool Function(TraceFrame frame)? filterFrame}) {
  final frames = Trace.from(StackTrace.current).frames;
  var index = 0;
  return frames
      .where((frame) => !frame.library.endsWith('/hook.dart'))
      .map((frame) => TraceFrame(
            index: index++,
            functionName: frame.member ?? '<anonymous>',
            file: frame.library,
            line: frame.line,
            column: frame.column,
            isAsync: false,
          ))
      .where((frame) => filterFrame?.call(frame) ?? true)
      .take(100)
      .toList()
      .asMap()
      .entries
      .map((entry) => TraceFrame(
            index: entry.key,
            functionName: entry.value.functionName,
            file: entry.value.file,
            line: entry.value.line,
            column: entry.value.column,
            isAsync: entry.value.isAsync,
          ))
      .toList();
}

JsonValue _jsonValueOrPlaceholder(Object? value) {
  try {
    final text = jsonEncode(value);
    return jsonDecode(text);
  } catch (_) {
    return const {'unavailable': 'not JSON serializable'};
  }
}

Map<String, Object?> _errorDetails(Object error) {
  if (error is Error || error is Exception) {
    return {
      'name': error.runtimeType.toString(),
      'message': error.toString(),
    };
  }
  return {'name': 'Error', 'message': error.toString()};
}

void _store(CreateHookOptions options, FunctionTrace trace) {
  final client = options.client;
  if (client == null) return;
  saveTrace(client, trace.toJson(),
          SaveTraceOptions(id: trace.traceId, ttlSeconds: options.ttlSeconds))
      .then((id) => options.onStored?.call(id))
      .catchError((Object error) =>
          options.onStoreError?.call(error, trace.traceId));
}

HookedFunction createHook(Function fn, [CreateHookOptions? options]) =>
    HookedFunction._(fn, options ?? const CreateHookOptions());
